User-guide - pandas : Add alternative to xarray.Dataset.from_dataframe

loco-philippe opened the pull request:
This PR follow the issue #9015 as proposed by @max-sixty.

I added an additional section in the pandas.rst file to provide a third-party pandas interface alternative that is lossless and reversible.

The main contribution is the ability to find the multidimensional structure hidden by the tabular structure.

max-sixty commented:
Overall I think this looks reasonable! I don't have strong views on the substance, others should comment if they have views...

We'll need to add `ntv_pandas` to the docs dependencies for the docs tests to pass.

loco-philippe commented:
@max-sixty 

Thanks for the answer to the question I was going to ask!

Would it also be useful to add `ntv_pandas` in the ecosystem.rst file (for example in the 'Extend xarray capabilities' category) ?

mathause commented:
> Would it also be useful to add `ntv_pandas` in the ecosystem.rst file (for example in the 'Extend xarray capabilities' category) ?

Yes, please go ahead.

mathause commented on doc/user-guide/pandas.rst:
```diff
@@ -110,6 +110,62 @@ work even if not the hierarchical index is not a full tensor product:
+
+__ https://github.com/loco-philippe/ntv-pandas/blob/main/README.md
+
+DataFrame to Dataset or DataArray
```
It's not clear how you would transform to a DataArray

loco-philippe replied:
I propose to add a Note: 

"a `dataset=True` or `dataset=False` parameter is used the choose the conversion result (active only if the Dataframe contains a single variable)"

mathause commented on doc/user-guide/pandas.rst:
```diff
@@ -110,6 +110,62 @@ work even if not the hierarchical index is not a full tensor product:
+Dataset or DataArray to Dataframe
+---------------------------------
+
+In the other direction, information that is not supported by the DataFrame must be transferred to the DataFrame (e.g. ``attrs`` data).
```
* Can you use `ds` from above (`ds = df.npd.to_xarray()`) to show the round trip?
* This sentence sounds like `attrs` is important for the structure, but it does not seem to be.
* Maybe you can start with the back-transformation from above and just mention the `attrs` business afterwards?

loco-philippe replied:
Yes you are right, `attrs` is not important for the structure. It's important only to be lossless.

I will follow your advice in the next commit.

mathause reviewed (commented):
Some comments. And as mentioned you will need to add the new dependency to https://github.com/pydata/xarray/blob/main/ci/requirements/doc.yml

loco-philippe commented on doc/user-guide/pandas.rst:
```diff
@@ -110,6 +110,62 @@ work even if not the hierarchical index is not a full tensor product:
+
+The conversion is done without loss, by finding the multidimensional structure hidden by the tabular structure.
+
+By applying this conversion to the DataFame above, we find the initial ``Dataset``:
```
The initial ``Dataset`` is the Dataset created at the beginning of the pandas.html page (section 'Dataset and DataFrame')

loco-philippe commented on doc/user-guide/pandas.rst:
```diff
@@ -110,6 +110,62 @@ work even if not the hierarchical index is not a full tensor product:
+
+.. ipython:: python
+
+    df = npd.from_xarray(ds)
```
Yes, you are right, with xarray accessors, it will be symmetrical (best solution).

I looked at the docs and I think this will be implemented in the next version.

loco-philippe commented:
Error Read the Docs : 

    ModuleNotFoundError                       Traceback (most recent call last)
    Cell In[11], line 1
    ----> 1 import ntv_pandas as npd
    
    ModuleNotFoundError: No module named 'ntv_pandas'
    
    <<<-------------------------------------------------------------------------

Is it because the package name (ntv-pandas) is different from the module name (ntv_pandas) ?

max-sixty commented:
> ModuleNotFoundError: No module named 'ntv_pandas'

Is it available on conda? Otherwise if only on `pip` then place it at the bottom in the `pip` section

mathause commented:
Now you get a syntax error because `arr[*idx]` is only available in python 3.11 while our docs are in python 3.10 (and your ntv-numpy package is python 3.9+ - you should add tests for your minimal python version and dependencies).

https://github.com/loco-philippe/ntv-numpy/blob/4b57b8cc1bfab749c01ddf7edbc38a9ef53623df/ntv_numpy/xconnector.py#L268-L269

mathause commented:
Checking again - you absolutely need to start testing your packages continuously. I am reluctant to 'endorse' a package that does not have a CI pipeline. Let us know if you need support with that.

loco-philippe commented:
@mathause

This bug is clearly unacceptable (I'm ashamed) !!

This demonstrates that I now need to take the time to have a robust CI process.
I will first use GitHub Actions to have a build and test pipeline (if you have minimum requirements to respect to endorse packages in Xarray or other advice, I'm interested!)

For the current PR, two solutions:
- solution 1: I correct the identified bug and I check that all the tests are validated on each of the python versions before making a new commit,
- solution 2: I integrate the above actions into a CI pipeline before making a new commit

It seems to me that solution 2 is preferable (unless you want to go quickly and in which case I will follow solution 1).

mathause commented:
No worries - I made a similar mistake before (CI pipeline not installing the correct python version/ environment). We have some guidelines for projects wanting to move to the `xarray-contrib` organisation, see https://github.com/xarray-contrib/xarray-contrib (not exactly the same but close enough).

For this PR I suggest to convert it to draft mode and we can revisit once you fix the issues. For the GHA pipelines - let me know (it depends a bit if you use conda or pip to install your packages).

keewis commented:
I wonder if, instead of describing code from a third-party package, we should add a "see also" or similar to the section? I.e. more prominent than an entry in the ecosystem page, but without putting actual code into the documentation. That way, we don't need to change anything if the third-party package changes (reducing our maintenance burden), and we also don't need to install even more things to the docs environment.

(To be clear, in posting this I'm trying to find general guidelines for what to include in the docs of `xarray` – it's definitely important to have this somewhere, my question is just where to put it, and how to make it easy to find)

loco-philippe commented:
@mathause 

For the GHA pipeline, I did some tests on the tabular structure analysis package (with pip) and it seems to work well ([ci.yml](https://github.com/loco-philippe/tab-analysis/blob/main/.github/workflows/ci.yml)): 
- tests are ok with 3.11,
- with 3.10, I find the problem arr[*idx]
- in 3.9 I also have 'match / case'

Initially, I will implement this simple solution (if I do not encounter dependency problems) with possibly some adaptations but my objective is then to move on to conda (are there any particular points to look at for that ?)

I will then look to integrate the packaging.

@keewis 

I understand the difficulty of both including third-party packages and reducing the maintenance burden.

I will therefore adapt to the choices you make on this subject.

loco-philippe commented:
The issue is fixed (new version of the `ntv_numpy` package) and the CI GHA pipeline integrates the tests with python versions 3.10 and 3.11.

I will then add xarray accessors as defined and build a new version.

mathause commented:
Cool great to hear!

Thinking about this - I would second @keewis idea. Featuring a prominent section without code would give it visibility, limit maintenance burden, and keep the docs environment smaller. (I maintain another smaller package where a third party package is featured in the docs. While I find it super cool that someone created an extension, it has caused above issues for me.)

---

Maybe this could be something along the lines:

```rst
Lossless and reversible conversion
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The previous example shows that the conversion is not reversible (lossy roundtrip) and that the size of the ``datasets`` increases. To avoid these problems, the third-party `ntv-pandas`__ library offers lossless and reversible conversions between ``Dataset``/ ``DataArray`` and pandas ``DataFrame`` objects.

__ https://github.com/loco-philippe/ntv-pandas
```

If you have not done so yet, you can showcase the examples in the docs of ntv-pandas.

loco-philippe commented:
The proposal to have Xarray documentation without code linked to a third party package seems logical to me.

So I modified the PR taking into account @mathause's proposal (with some modifications).

I also added the Xarray accessors in the ntv_numpy package. We now have symmetric methods `Dataset.nxr.to_dataframe` and `DataFrame.npd.to_xarray`).

Another question: I haven't found any other tools or methodologies that analyze the structure of a tabular data to extract the multidimensional structure. Do you know any?

mathause commented on doc/user-guide/pandas.rst:
```diff
@@ -110,6 +110,29 @@ work even if not the hierarchical index is not a full tensor product:
+- a non-dimension Dataset ``coordinate`` is converted into ``variable``
+- a non-dimension DataArray ``coordinate`` is not converted
+- ``dtype`` is not allways the same (e.g. "str" is converted to "object")
+- ``attrs`` metadata is not converted
```
```suggestion
- ``attrs`` metadata is not conserved
```
?

mathause commented on doc/user-guide/pandas.rst:
```diff
@@ -110,6 +110,29 @@ work even if not the hierarchical index is not a full tensor product:
+
+The `ntv-pandas examples`__ show how to improve the conversion for the previous ``Dataset`` example and for more complex examples.
+
+__ https://github.com/loco-philippe/ntv-pandas
+__ https://github.com/loco-philippe/ntv-pandas/tree/main/example
+
```
```suggestion
```

It's maybe better to inline these links, especially now that there is more than one.

mathause commented on doc/user-guide/pandas.rst:
```diff
@@ -110,6 +110,29 @@ work even if not the hierarchical index is not a full tensor product:
+- ``dtype`` is not allways the same (e.g. "str" is converted to "object")
+- ``attrs`` metadata is not converted
+
+To avoid these problems, the third-party `ntv-pandas`__ library offers lossless and reversible conversions between
```
```suggestion
To avoid these problems, the third-party `ntv-pandas <https://github.com/loco-philippe/ntv-pandas>`__ library offers lossless and reversible conversions between
```

mathause commented on doc/user-guide/pandas.rst:
```diff
@@ -110,6 +110,29 @@ work even if not the hierarchical index is not a full tensor product:
+
+This solution is particularly interesting for converting any ``DataFrame`` into a ``Dataset`` (the converter find the multidimensional structure hidden by the tabular structure).
+
+The `ntv-pandas examples`__ show how to improve the conversion for the previous ``Dataset`` example and for more complex examples.
```
```suggestion
The `ntv-pandas examples <https://github.com/loco-philippe/ntv-pandas/tree/main/example>`__ show how to improve the conversion for the previous ``Dataset`` example and for more complex examples.
```

mathause replied:
Do you have these examples in the ntv-pandas docs somewhere?

loco-philippe replied:
The link of a more complex structure is in the first section of the Notebook (and in the [example README](https://github.com/loco-philippe/ntv-pandas/tree/main/example)).

This [example](https://nbviewer.org/github/loco-philippe/ntv-pandas/blob/main/example/example_multidimensional.ipynb) is not dedicated to Xarray but if you think it would be better to have the both examples in the same Xarray example, i can add the cells in the Notebook.

If you also have another example, I can add that too.

mathause commented on ci/requirements/doc.yml:
```diff
@@ -44,3 +44,5 @@ dependencies:
   - pip:
       # relative to this file. Needs to be editable to be accepted.
       - -e ../..
+      - ntv-pandas>=2.0.0
+      - tabulate>=0.9.0
```
```suggestion
```

mathause reviewed (changes requested):
Thanks for implementing our suggestions. Need to remove the packages from the doc.yml again. I have some optional minor suggestions, but looks good to go.

mathause reviewed (approved):
Looks good!

loco-philippe commented:
Thank you @max-sixty and @mathause for your approval of this PR.

I finally added another [example (use case)](https://nbviewer.org/github/loco-philippe/ntv-pandas/blob/main/example/example_xarray.ipynb) which I think should address new needs (but I don't know Xarray well enough and I'm interested in your opinion).

Note: JSON, Pandas and Xarray interfaces are built on a neutral format as defined in [this notebook](https://nbviewer.org/github/loco-philippe/ntv-numpy/blob/main/example/example_ntv_numpy.ipynb). I think this structure is consistent with your roadmap (I'm also interested in your opinion).

mathause commented:
close/ open to trigger the tests again

mathause commented:
No idea why this does not work, but I don't see this could have something to do with the PR itself. I'll merge manually.

Thanks for your PR and willingness to adopt to our changes!
